TechnicalDocuments/Proposals/IdentifiersForUIAutomation.md

```diff
@@ -0,0 +1,203 @@
+    func updateAccessibilityIdentifiers(usingPrefix prefix: String) {
+        return prefix + automationName
+    }
+}
+```
+
+and this particular composite component called `prettyCard`:
+
+```swift
+extension DesignLibrary.ComponentsLibrary {
+    public func prettyCard() -> AnyRenderable {
+        return [
+            UI.Label(/* ... */, automationName: String(AutomationName.prettyCardNotice)),
+            UI.Button(/* ... */, automationName: String(AutomationName.prettyCardSubmit))
+        ].stack(axis: .vertical)
+    }
+}
+
+extension DesignLibrary.AutomationName {
+    public enum PrettyCard: String {
+        case notice
+        case submit
+    }
+}
+```
+
+When we render `prettyCard`:
+
+```
+render(
+    Box(
+        sections: [
+            Section(id: .activity, items: [
+                Node(id: .card, component:
+                    prettyCard()
+                )
+            ])
+        ]
+    )
+)
+```
+
+It should result in two accessibility identifiers:
+
+* `activity/card/notice` applied to the label; and
+* `activity/card/submit` applied to the button.
+
+Notice that automation names for the relevant subcomponents of `prettyCard` has been published in the `DesignLibrary.AutomationName` umbrella. This facilitates the next suggestion of this proposal...
+
+### Working with UI automation tests
+
+Now that we have a systematic approach of defining and publishing information about accessibility identifiers, we can use this to reliably generate accessibility identifiers in UI automation tests!
+
+The proposal suggests that the UI automation test target should link with all UI frameworks. This allows the automation code to have access to `Renderer`s and any published `AutomationName` enums, and can therefore construct accessibility identifiers. Changes made to any section ID, item ID or automation name enums would either be automatically picked up by the automation code, or lead to compile-time breakage depending on the circusmtances.
+
+`BaseScreen` shall be extended to provide a conveinence:
+```swift
+extension BaseScreen {
+    subscript<Renderer: BoxRenderer>(
+        _ renderer: Renderer.Type,
+        _ section: Renderer.SectionID,
+        _ item: Renderer.ItemID
+    ) -> String {
+        return "\(section)/\(item)/"
+    }
+
+    subscript<Renderer: BoxRenderer, AutomationName: RawRepresentable>(
+        _ renderer: Renderer.Type,
+        _ section: Renderer.SectionID,
+        _ item: Renderer.ItemID,
+        _ automationName: AutomationName
+    ) -> String where AutomationName.RawValue == String {
+        return "\(section)/\(item)/" + automationName.rawValue
+    }
+}
+```
+
+which the automation code can rely on. For example:
+
+```swift
+import BabylonHealthManagementUI
+
+extension BaseScreen {
+    // NOTE: alias in the UI testing target to help reduce verbosity.
+    typealias Name = DesignLibrary.AutomationName
+}
+
+class FeedScreen: BaseScreen {
+    func tapSubmit() {
```

Comment:
I'm probably missing something obvious, but were is FeedRenderer being defined? and how does the screen object know which renderer.self to invoke

Reply:
`FeedRenderer` is defined in the UI frameworks or the main app target, which is the actual entity who maps screen states into components for rendering.